json-schemas/HOL-PART3.md

```diff
@@ -0,0 +1,162 @@
+# Part 3: Array Validation
+
+With our newly learned knowledge of validation for sub-documents, let's
+go a step further and look at how to validate another feature of
+MongoDB's JSON Schema: arrays.
+
+Our use case here will be the modeling of a cooking recipe. Typically a
+recipe will have a list of ingredients. In a `recipe` collection then,
+our documents would have, among other things, an array of ingredients.
+Let's assume that for each ingredient listed we need a *quantity*,
+*measurement*, *ingredient*, and an optional *format*.
+
+1. Let's define a *basic* recipe skeleton:
+```
+    db.recipes.drop()
+    db.createCollection ( "recipes",
+    {
+        validator:
+                    {
+         $jsonSchema:
+    {
+        bsonType: "object",
+        required: ["name", "ingredients"],
+        properties:
+        {
+            _id: {},
+            name: {
+                bsonType: "string",
+                description: "'name' must be a string and is required"
+                  },
+            ingredients: {
+                bsonType: ["array"],
+                minItems: 1, // each recipe must have at least one ingredient
+                items: {
+                    bsonType: ["object"],
+                    required: ["quantity", "measurement", "ingredient"],
+                    additionalProperties: false,
+                    description: "'ingredients' must contain the stated fields.",
+                    properties: {
+                    quantity: {
+                      bsonType: ["double", "decimal"],
+                      description: "'quantity' is required and is of double or decimal type"
+                            },
+                    measurement: {
+                      enum: ["tsp", "Tbsp", "cup", "ounce", "pound",  "each"],
+                      description: "'measure' is required and can only be one of the given enum values"
+                            },
+                    ingredient: {
+                      bsonType: "string",
+                      description: "'ingredient_name' is required and is a string"
+                            },
+                    format: {
+                      bsonType: "string",
+                      description: "'format' is an optional field of type string"
+                            }
+                    }
+                }
+            }
+
+                }
+            }
+        }
+    }
+    )
+```
+
```

Comment:
Isn't it worth mentioned the `array` bsonType value and the `items` property, and explain how that creates a schema for array items?  Maybe also adding a pointer to https://tools.ietf.org/html/draft-fge-json-schema-validation-00#page-9 would be worth?